src/hammer-vlsi/hammer_utils/__init__.py

```diff
@@ -101,6 +97,75 @@ def in_place_unique(items: List[Any]) -> None:
             i += 1
 
 
+def coerce_to_grid(num: float, grid: Decimal) -> Decimal:
+    """
+    Coerce a floating-point number to the nearest multiple of the provided grid
+
+    :param num: The input floating-point number
+    :param grid: The decimal grid value to which num should be coerced
+    :return: A decimal number on-grid
+    """
+    return Decimal(round(num / float(grid))) * grid
+
+def check_on_grid(num: Union[float, Decimal], grid: Decimal) -> bool:
+    """
+    Checks if a number is an integer multiple of a specified grid unit.
+    This supports floats and Decimals, although support for floats will eventually be removed.
+    Removal of float support will be in ucb-bar/hammer#376
+
+    :param num: The number to check
+    :param grid: The decimal grid value on which to check number
+    :return: True if num is on-grid, False otherwise
+    """
+    # TODO(johnwright) Remove float support in ucb-bar/hammer#376
+    num_dec = num if isinstance(num, Decimal) else coerce_to_grid(num, Decimal("0.001"))
+    assert isinstance(num_dec, Decimal)
+    return Decimal(int(num_dec / grid)) == (num_dec / grid)
+
+def gcd(*values: int) -> int:
+    """
+    Return the greatest common divisor of a series of ints
+
+    :param values: The values of which to compute the GCD
+    :return: The GCD
+    """
+    assert len(values) > 0
+    if len(values) == 1:
+        return values[0]
+    elif len(values) == 2:
```

Comment:
There is a math.gcd, so maybe use that in the base case here. It seems like a fold operation would work here for the whole thing but not familiar with the python functools stuff.

Reply:
Ditto above. Python `reduce()` can double for Scala `reduce` and `fold`

Reply:
OK, I can make it more functional

Reply:
done